### PR TITLE
Damaris integ v3

### PR DESCRIPTION
**Pull request by ElyesAhmed**
This is a draft PR to integrate Damaris Library to OPM. The current test is only used to output "PRESSURE".

### CODE REVIEW

**Comment by atgeirr** [comment]
@akva2 Can you take a look at the build system side of this?

File: ebos/eclwriter.hh [hunk 1]
```diff
@@ -39,8 +39,22 @@
...
+#include <Damaris.h>
+#endif
+
+#if OPM_HAVE_DAMARIS
+#include <Damaris.h>
+#endif
+
+#if HAVE_DAMARIS
+#include <Damaris.h>
+#endif
```
Comment (akva2) [inline ebos/eclwriter.hh:48]: i'd like to see this a fifth time so we are REALLY sure ;P

and don't remove the other includes.

File: opm-simulators-prereqs.cmake [hunk 2]
```diff
@@ -17,7 +17,7 @@ set (opm-simulators_CONFIG_VAR
   DUNE_ISTL_VERSION_MAJOR
   DUNE_ISTL_VERSION_MINOR
   DUNE_ISTL_VERSION_REVISION
-  HAVE_SUITESPARSE_UMFPACK
+  HAVE_SUITESPARSE_UMFPACK
```
Comment (akva2) [inline opm-simulators-prereqs.cmake:20]: no.

File: CMakeLists.txt [hunk 3]
```diff
@@ -542,6 +549,38 @@ if(HAVE_FPGA)
   ExternalProject_Get_Property(FPGA_library binary_dir)
   target_link_libraries(opmsimulators PUBLIC ${binary_dir}/fpga_lib_alveo_u280.a)
 endif()
+if (USE_DAMARIS_LIB) # User requested Damaris support
+        if (Damaris_FOUND)
```
Comment (akva2) [inline CMakeLists.txt:554]: please make a wrapper find module instead of putting all of this in the main cmakelists.

File: opm/simulators/flow/Main.hpp [hunk 4]
```diff
@@ -197,18 +210,38 @@ class Main
 #endif
         EclGenericVanguard::setCommunication(std::make_unique<Parallel::Communication>());
 
-        handleTestSplitCommunicatorCmdLine_();
+/* Damaris I/O initialization has been moved to after the readDeck() call in Main::initalize_()
```
Comment (akva2) [inline opm/simulators/flow/Main.hpp]: no new commented code. add it if/when it is required.

File: opm/simulators/utils/readDeck.cpp [hunk 5]
```diff
@@ -266,9 +266,9 @@ namespace {
                                 *parseContext, errorGuard);
         }
 
-        if (eclipseState == nullptr) {
+        //if (eclipseState == nullptr) {
```
Comment (akva2) [inline opm/simulators/utils/readDeck.cpp]: if we don't want the if, remove it.

File: opm/simulators/utils/readDeck.cpp [hunk 6]
```diff
@@ -309,9 +309,9 @@ namespace {
                                        std::unique_ptr<Opm::WellTestState>& wtestState,
                                        std::shared_ptr<Opm::SummaryConfig>& summaryConfig)
     {
-        if (eclipseState == nullptr) {
+       // if (eclipseState == nullptr) {
```
Comment (akva2) [inline opm/simulators/utils/readDeck.cpp]: ditto.

File: opm/simulators/utils/readDeck.cpp [hunk 7]
```diff
@@ -477,7 +477,7 @@ Opm::setupLogging(const int          mpi_rank_,
     return output;
 }
 
-void Opm::readDeck(Opm::Parallel::Communication    comm,
+void Opm::readDeck(Opm::Parallel::Communication&    comm,
```
Comment (akva2) [inline opm/simulators/utils/readDeck.cpp]: seems unrelated. these structs are very cheap to copy so that's been the preferred semantics.

**Review by akva2 — COMMENTED** [review]
mostly commented on buildsystem, but also some other stuff i noticed.

**Comment by alfbr** [comment]
I assume Damaris is a new dependency?

**Comment by akva2** [comment]
it's an optional dependency.

**Comment by alfbr** [comment]
> it's an optional dependency.

Sounds good. Has anybody considered the implications of deploying it? I am thinking about existence of rpms for Red Hat or debs for Ubuntu/Debian. In addition. is the code base well maintained?

**Comment by bska** [comment]
> but also some other stuff i noticed.

@akva2: Those comments **mostly** appear to apply to `readDeck.cpp` which isn't touched at all in this PR.  Did GitHub bring in a stale ref before?

**Comment by akva2** [comment]
I am seem to have accidentially clicked the commit listed in the commit message of the second commit.

**Comment by ElyesAhmed** [comment]
> I am seem to have accidentially clicked the commit listed in the commit message of the second commit.

Yes Arne Morten, its two commits. However,  the second commit is still not a good set of commits with unclear messages. I wanted to preserve the authorship because Joshua Bowden from INRIA also contributed to this work! Maybe you have an idea on how to clean it under this constraint!

**Comment by akva2** [comment]
you can safely split it, and use git commit --author to preserve authorship.

**Comment by bska** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
The flaws mentioned previously are fixed, and I have made modifications to the code that I thought were necessary. I think this is ready to be merged, except for @akva2's request to put the Damaris build system stuff in its own function. I have not done this, as I am not very CMake-skilled, but I assume it is not so hard?

**Comment by atgeirr** [comment]
jenkins build this serial please

**Comment by akva2** [comment]
```
CMake Error at opm-common/cmake/Modules/OpmCompile.cmake:41 (add_library):
  Cannot find source file:

    /home/akva/kode/opm/opm-simulators/opm/simulators/utils/initDamarisXMLFile.cpp
```
XML -> Xml

**Comment by ElyesAhmed** [comment]
I think we fixed all issues raised by reviewers now, including those related to the build-system

**Comment by bska** [comment]
jenkins build this serial please

**Comment by atgeirr** [comment]
jenkins build this serial please

**Comment by atgeirr** [comment]
All green, merging. Thanks for contributing, reviewing and commenting!

**Comment by ElyesAhmed** [comment]
> All green, merging. Thanks for contributing, reviewing and commenting!

thank you all!

**Comment by blattms** [comment]
mmh, is this tested with DUNE 2.6?

**Comment by atgeirr** [comment]
> mmh, is this tested with DUNE 2.6?

I assumed it was ok since Jenkins was green, but we may have to use a special trigger for that? In any case, do you see any problems? Would be a bit surprising, since this is not really touching on Dune things.

File: ebos/eclwriter.hh [hunk 8]
```diff
@@ -281,10 +302,30 @@ public:
...
+            }
 
+            if (!isSubStep) {
+                data::Solution localCellData = {};
+                this->eclOutputModule_->assignToSolution(localCellData);
+                // Output the PRESSURE field
+                if (this->eclOutputModule_->getPRESSURE_ptr() != nullptr) {
+                    damaris_write("PRESSURE", (void*)this->eclOutputModule_->getPRESSURE_ptr());
+                    damaris_end_iteration();
+                }
```
Comment (bska) [inline ebos/eclwriter.hh:325]: I didn't pick up on this before, but this sequence of statements risks reading from moved-from memory.  The reason is that [`assignToSolution()`](https://github.com/OPM/opm-simulators/blob/2064e8725f11a5ae00e0fd96a686809fd6eab983/ebos/eclgenericoutputblackoilmodule.cc#L645) currently [transfers](https://github.com/OPM/opm-simulators/blob/2064e8725f11a5ae00e0fd96a686809fd6eab983/ebos/eclgenericoutputblackoilmodule.cc#L655) ownership of its solution arrays into the `data::Solution` object, including that of the array that holds [`PRESSURE`](https://github.com/OPM/opm-simulators/blob/2064e8725f11a5ae00e0fd96a686809fd6eab983/ebos/eclgenericoutputblackoilmodule.cc#L683) values.  There is also no guarantee that `vector::data()` is `nullptr` after it has been moved from.

Furthermore, by calling `assignToSolution()` and just discarding the `localCellData`, you remove all the data arrays that would otherwise go to the ECLIPSE-compatible restart file.  Is that really intentional?  My suggestion here would be to remove the call to `assignToSolution()` here and just use the `PRESSURE_ptr()` functions exclusively.

**Comment by blattms** [comment]
> > mmh, is this tested with DUNE 2.6?
> 
> I assumed it was ok since Jenkins was green, but we may have to use a special trigger for that? In any case, do you see any problems? Would be a bit surprising, since this is not really touching on Dune things.

Jenkins only checks 2.7. If you look at some of my current PRs (DUNE 2.9), you will notice that some of those also fix 2.6. I am just a bit anxious, this near to the release...

**Comment by atgeirr** [comment]
> Jenkins only checks 2.7. If you look at some of my current PRs (DUNE 2.9), you will notice that some of those also fix 2.6. I am just a bit anxious, this near to the release...

Thanks for the info. I assume that we can rerun Jenkins on those PRs to know if we did something wrong here? If so, I will prioritize fixing it.

File: ebos/eclwriter.hh (hunk 8 above)
Comment (atgeirr) [reply ebos/eclwriter.hh:325]: Thanks for the explanation. I agree that the assignToSolution() is incorrect here. The PRESSURE_ptr() is a temporary API for testing, I think long term we should get the output arrays from the data::Solution object. Then it may be correct to reinstate it, or rather to make sure we only do assignToSolution() once by unifying with the parts further down in the code.

I will test and push a PR to remove the assignToSolution() call now.
